### PR TITLE
Combine direct_split with TCB Desynchronization

### PR DESCRIPTION
**Pull request by Lanius-collaris**
Related to #46

Todo:

- [X] Add more comments
- [ ] Discuss functionality of this dialer.

### CODE REVIEW

**Comment by ignoramous** [comment]
Thanks; how are you building and testing this? Or, is it PoC (which is also fine)?

**Comment by Lanius-collaris** [comment]
[PoC.patch.txt](https://github.com/celzero/firestack/files/14968234/PoC.patch.txt)
@ignoramous You can use this patch. Build and run it with adb or termux.

**Comment by Lanius-collaris** [comment]
@ignoramous Does firestack have API(s) to configure dialers?

**Comment by ignoramous** [comment]
> Does firestack have API(s) to configure dialers?

Yes; see `[r|n|p|tls]dial.go` in [`intra/dialers`](https://github.com/celzero/firestack/tree/n2/intra/dialers). Not the most stellar code.

**Comment by Lanius-collaris** [comment]
@ignoramous Have you design how rethink-app side specifies a strategy? We are adding multiple circumvention strategies. Can we use AddProxy()? (Sorry, I'm not familiar with rethink-app)
And dialers.SplitDial() may trigger residual censorship, I think we can make WeirdSplit the default strategy. (It doesn't break any load balancer but increases latency, cause servers to wait for retranssmission)

https://github.com/celzero/firestack/blob/e16959c082c3450bdb8d826da0285b787b315491/intra/dialers/rdial.go#L88-L91
Can we remove this filter?

**Comment by ignoramous** [comment]
> Can we use AddProxy()? (Sorry, I'm not familiar with rethink-app)

Don't have to. `ipn/exit.go` (no fancy stuff, just plain old egress), `ipn/base.go` (splits `ClientHello` if TCP on ports `443` & `853`), and `ipn/ground.go` (drops all egress) are pre-added "proxies" that use the underlying network as-is. In this case, we probably switch `ipn/base.go` to use `WeirdDial` (needs a better name) if you think it is better than `SplitDial` in all scenarios, with a fallback to `SplitDial` (but that can be implemented in a later commit).

**Comment by Lanius-collaris** [comment]
[overwrite_split_experiment branch](https://github.com/Lanius-collaris/firestack/tree/overwrite_split_experiment)
@ignoramous  Could you try it? Please run it on hardware devices, it doesn't work with SLIRP.

**Comment by Lanius-collaris** [comment]
@ignoramous Do you have time to review my pull request next week?

**Comment by Lanius-collaris** [comment]
Doesn't work on main branch.

**Comment by ignoramous** [comment]
> Do you have time to review my pull request next week

I've reviewed this and nothing looks particularly concerning to me. There's a few linter remarks on the code, which could be addressed.

**Comment by Lanius-collaris** [comment]
Received ICMP messages may not contain my payload, so the implementation of traceroute needs modification.

@ignoramous `SwitchStrategy()` doesn't work. Move it to `ipn/exit.go`?

**Comment by ignoramous** [comment]
> SwitchStrategy() doesn't work. Move it to ipn/exit.go?

Could instead move var `DialStrategy` to `config/settings.go` as `atomic.Int32`, and the method `DialStrat(int32)` to `android/tun2socks.go`.

File: intra/dialers/split_and_desync.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+	} else {
+		err = unix.SetsockoptInt(udpFD, unix.IPPROTO_IP, unix.IP_RECVERR, 1)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var msgBuf [probeSize]byte
+	var ttl int
+	basePort := 1 + randNum.Intn(65535-maxTTL)
```
[the remark this answers is not in the document]
Comment (Lanius-collaris) [reply intra/dialers/split_and_desync.go]: 🙃

**Comment by Lanius-collaris** [comment]
It's usable now. What's next? @ignoramous

File: intra/dialers/split_and_desync.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+	if err != nil {
+		return 0, err
+	}
+	var offset int64 = 0
+	n1, err := unix.Sendfile(sockFD, fileFD, &offset, len(s.Payload))
+	if err != nil {
+		return n1, err
+	}
+
+	copy(firstSegment, b[:len(s.Payload)])
```
Comment (ignoramous) [inline intra/dialers/split_and_desync.go:319]: @Lanius-collaris what writes to conn this newly copied to `firstSegment`? Or, is it not required and written to in a retransmission (somehow...)?

**Pull request merged**